ompi/util/timings.h

```diff
@@ -4,123 +4,215 @@
+                free(tmp->item->desc);                                         \
+                tmp->item->desc = NULL;                                        \
+            }                                                                  \
+            if ( tmp->item->in ) { free(tmp->item->in); }                      \
+            free(tmp->item);                                                   \
+            tmp->item = NULL;                                                  \
+        }                                                                      \
+        free(tmp);                                                             \
+        tmp = NULL;                                                            \
+    } while ( NULL != t);                                                      \
+exit:;                                                                         \
+    OMPI_TIMING_env.cnt = 0;                                                   \
+})
+
+#define OMPI_TIMING_INIT(_inum)                                                \
+    ompi_timing_env_t OMPI_TIMING_env;                                         \
+    OMPI_TIMING_env.prefix = __FUNCTION__;                                     \
+    OMPI_TIMING_env.inum = _inum;                                              \
+    OMPI_TIMING_env.get_ts = opal_timing_ts_func(OPAL_TIMING_AUTOMATIC_TIMER); \
+    OMPI_TIMING_env.cnt = 0;                                                   \
+    OMPI_TIMING_env.error = 0;                                                 \
+    OMPI_TIMING_env.ts = OMPI_TIMING_env.get_ts();                             \
+    OMPI_TIMING_env.timing = (ompi_timing_list_t*)malloc(sizeof(ompi_timing_list_t));       \
+    OMPI_TIMING_env.timing->next = NULL;                                       \
+    OMPI_TIMING_ITEM_INIT(&OMPI_TIMING_env.timing->item);                      \
+
+#define OMPI_TIMING_NEXT(fmt, ...) {                                           \
+    ompi_timing_item_t *item;                                                  \
+    if( OMPI_TIMING_env.cnt &&                                                 \
+                !(OMPI_TIMING_env.cnt % OMPI_TIMING_env.inum) ){               \
+        OMPI_TIMING_ENV_ITEM_EXTEND;                                           \
+    }                                                                          \
+    item = OMPI_TIMIG_ENV_ITEM_IDX(OMPI_TIMING_env.cnt);                       \
+    if ( NULL == item ) {                                                      \
+        printf("OMPI_TIMING [%d %s]: item not found\n",                        \
+            __LINE__, __FUNCTION__);                                           \
+        abort();                                                               \
+    }                                                                          \
+    item->in[OMPI_TIMING_env.cnt % OMPI_TIMING_env.inum] =                     \
```

Comment:
I don't want extra memory allocation.
I want description field to be of fixed size, but we check for overflow. as it was directed in the draft.